Handle missing resources during upgrade and rollback

fmuyassarov opened the pull request:
Updated logic to check if a resource exists in the cluster.

Attempt to fix: https://github.com/helm/helm/issues/13437 

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

fmuyassarov commented:
/cc @mattfarina @matthiasvegh

fmuyassarov commented:
@mattfarina Do you see any potential consequences of this change? I tested it with `--atomic` upgrade and rollback, and it seems to help prevent false failures (e.g., when a resource does exist in the cluster).

matthiasvegh commented:
@fmuyassarov I've not yet been able to do testing on this yet, I'll be able to later this week. At first glance you're skipping the update of any manifests that are present in the target but not in the source if they also exist in the cluster.
While this is okay in the trivial case where an upgrade fails before removing a manifest not present in the target and then rolling back to where it is present, it fails for more involved situations.

As an example, if a Helm chart contains a TLS certificate as a Secret in revision 1, but in revision 2 has a cert-manager.io/v1 Certificate with the same name to indirectly create that Secret. The Secret is of course no longer part of the release but is still conceptually part of it as we expect the Cert Manager controller to create/update it. While the chart author may expect that Secret to work just as well in that chart, it's likely not exactly the same (if nothing else, helm.sh annotations are likely to be absent).

Skipping the update not only does not take you back to the previous state, any deviations will then be carried forward as out-of-band changes via 3-way merge.

robertsirc reviewed (changes requested):
Hello, first thank you for your pull request we do appreciate it. If we could ask can we please get some test around this change? I know there might not be current test but we are trying to ensure we don't break current functionality and up our test coverage.

fmuyassarov commented:
> Hello, first thank you for your pull request we do appreciate it. If we could ask can we please get some test around this change? I know there might not be current test but we are trying to ensure we don't break current functionality and up our test coverage.

Hi @robertsirc . Sure on it, will add a separate commit with tests.

fmuyassarov committed:
Handle missing resources during upgrade and rollback

Check if a resource exists in the cluster before returning an error.
If the resource exists, the continue to updating thr resource if not,
return error.

Signed-off-by: Feruzjon Muyassarov <feruzjon.muyassarov@est.tech>

fmuyassarov commented:
@robertsirc PTAL. Updated the PR as per review from @matthiasvegh and realized that there is unit test coverage already in place for the Update function which I'm touching on this patch and I just had to adjust that test function little bit. I also checked the coverage percentage to make sure that it doesn't go below what it is currently (63%).

matthiasvegh commented:
Hi @fmuyassarov, I've done some testing, and while the spurious failure is gone, there is still no update made to the resource absent in the source.

Given the chart in the original issue #13437, I did:

  1. `helm install test .`
  2. `helm upgrade test .` (which fails in pre-, as expected)
  3. `kubectl edit configmap/cfg` and changed `data.foo` to `baz` (was `bar`)
  4. `helm rollback test 1` this did not fail, but left `data.foo` as `baz`.

For testing purposes, I'd suggest tweaking the ConfigMap to update `data.foo` itself in revision 2 before being deleted in revision 3, then rollback directly to revision 1 instead of just to 2. As things stand, this change only works for `--rollback` (and when nobody else in the release takes over the manifest). For convenience, I've wrapped this up in a Chart below.
<details>
<summary>Updated Chart to show lack of an update</summary>

```yaml
{{- if lt .Release.Revision 3 -}}
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: cfg
data:
  foo: bar-from-{{ .Release.Revision | toString }}
{{ end -}}

{{- if ge .Release.Revision 3 -}}
---
apiVersion: batch/v1
kind: Job
metadata:
  name: "failurecauser"
  annotations:
    helm.sh/hook: pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded,hook-failed
spec:
  template:
    spec:
      terminationGracePeriodSeconds: 1
      containers:
        - name: never
          image: "busybox:latest"
          args:
            - "/bin/sh"
            - "-c"
            - "false"
          imagePullPolicy: Always
      restartPolicy: Never
  backoffLimit: 0
{{ end -}}
```
</details>

With this chart, `helm install` results in `.data.foo` being `bar-from-r1`, `helm upgrade` results in `bar-from-r2`, the third `helm upgrade` fails, and a `helm rollback 1` results in `bar-from-r2` instead of `bar-from-r1`.

I'm uncertain on which direction to take for this fix. I can see two strategies being viable.
Either behave like `--force` by passing force=true and changing the log call to to not log the original `kind` in
https://github.com/helm/helm/blob/e3dffbd9ec4a54e98dc7517bc5f9145ba80ef07c/pkg/kube/client.go#L687 as I can't see a case where the kind changes, but maybe it depends on the `apiVersion`.
This results in the manifest being completely overwritten when reintroduced during a rollback. This seems like sane behavior, but gives Helm an edge-case where 3-way merge behaves differently to normal.

Alternatively, Helm could keep better track of which manifests were patched prior to the upgrade failing e.g., by adding a new annotation so you can compute a different source per manifest. This would result in Helm realizing that the manifest had not made it to the higher revision and use the older revision as the source, with 3-way merge preserving out-of-band changes as expected. I imagine this working by `Update` taking in _two_ previous releases, zipping them before iterating and determining the "original" based on whether they were ever actually patched. This seems like significant rework, and I don't want to recommend it as a mere user, I hope @robertsirc or @mattfarina can chime in.

fmuyassarov commented:
Hi @robertsirc,
I've reworked the patch to ensure that resources are updated correctly during a rollback. I verified this behavior using the example that @matthiasvegh pointed out earlier. One notable change I made after @matthiasvegh's tests and feedback is that I've enforced resource replacement during the `Update()` call by setting `force` flag to true.
Could you please take a look ?

fmuyassarov commented:
/cc @mattfarina

fmuyassarov commented:
/hold

robertsirc commented:
Hello, got time to look at this but your branch has some conflicts in it.

fmuyassarov commented:
Let's continue this fix on a new PR: https://github.com/helm/helm/pull/31412

/close
